src/v/resource_mgmt/available_memory.cc

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "resource_mgmt/available_memory.h"
+
+#include "prometheus/prometheus_sanitize.h"
+#include "seastarx.h"
+
+#include <seastar/core/memory.hh>
+#include <seastar/core/metrics.hh>
+
+#include <fmt/core.h>
+
+#include <iterator>
+#include <memory>
+#include <numeric>
+#include <stdexcept>
+
+namespace resources {
+
+deregister_holder::~deregister_holder() {
+    if (handle) {
+        parent->deregister(handle);
+    }
+}
+
+deregister_holder
+available_memory::inner_register(const ss::sstring& name, afn&& fn) {
+    auto handle = _next_handle++;
+    _reporters.push_back({name, std::move(fn), handle});
+    return deregister_holder{this, handle};
```

Comment:
should we make `available_memory` not movable and non copyable to prevent captured `this` from being changed ?

Reply:
Though this use of `this` has gone way thanks to your intrusive list suggestion, this still make sense so I've done it in the next push.